src/rust/engine/fs/src/lib.rs

```diff
@@ -388,7 +388,7 @@ pub enum SymlinkBehavior {
   Oblivious,
 }
 
-#[derive(Debug)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
```

Comment:
I'm surprised to see `Eq` implemented - I was reading that that's barely implemented.

Out of curiosity, how do you determine when you do and do not do it? I've read about the difference, but am curious if you have any heuristics you rely on.

Reply:
This is a potentially large object: you should probably not implement `Eq`/`Hash` on it. The reason `Key` was used in this position is that it is much, much smaller (just an integer).

Reply:
> Out of curiosity, how do you determine when you do and do not do it? I've read about the difference, but am curious if you have any heuristics you rely on.

In this case it was because `Snapshot` already implemented `Eq`, and I didn't dig into why that's the case; maybe `PartialEq` would be enough. The distinction between `PartialEq` and `Eq` in Rust basically exists in order to support the floating point number types; IEEE 754 specifies that floating-point numbers must satisfy `NaN != NaN`, which is inconsistent with the requirement on the [Eq trait](https://doc.rust-lang.org/std/cmp/trait.Eq.html) that `a == a` for all `a`, which `PartialEq` doesn't have. Since these types aren't floating point numbers and don't have that weird quirk of floating point numbers, I don't think the distinction matters for anything we're doing.